File: backend/src/tags/dialogue.py
import re
from src.tags.base import BaseTag, TagParseResult
from src.text_utils import clean_dialogue_text
from src.core import Character


class VariableSpeakerTag(BaseTag):
    """Handles :var: > "text" — speaker resolved from the runtime env."""
    name = "dialogue.variable"
    PATTERN = re.compile(r'^:([A-Za-z_][A-Za-z0-9_]*):\s*>\s*(.*)$')

    def parse(self, line, line_idx, ctx):
        m = self.PATTERN.match(line)
        if not m:
            return TagParseResult(consumed=False)
        return TagParseResult(
            step={
                "type": "dialogue",
                "speaker_mode": "variable",
                "key": m.group(1),
                "raw_text": clean_dialogue_text(m.group(2)), # Ensured field consistency
            },
            consumed=True,
        )

    def execute(self, step, ctx):
        if step.get("type") != "dialogue" or step.get("speaker_mode") != "variable":
            return None
        env = ctx["env"]
        key = step["key"]
        if key in env and isinstance(env[key], Character):
            name = env[key].name
        else:
            name = key
        return ("frame", _build_frame(step, name, ctx))


class LiteralSpeakerTag(BaseTag):
    """Handles Name > "text" — literal speaker name."""
    name = "dialogue.literal"
    PATTERN = re.compile(r'^([^>]+)>\s*(.*)$')

    def parse(self, line, line_idx, ctx):
        if line.startswith(">"):
            return TagParseResult(consumed=False)
        m = self.PATTERN.match(line)
        if not m:
            return TagParseResult(consumed=False)
        return TagParseResult(
            step={
                "type": "dialogue",
                "speaker_mode": "literal",
                "name": m.group(1).strip(),
                "raw_text": clean_dialogue_text(m.group(2)), # Ensured field consistency
            },
            consumed=True,
        )

    def execute(self, step, ctx):
        if step.get("type") != "dialogue" or step.get("speaker_mode") != "literal":
            return None
        return ("frame", _build_frame(step, step["name"], ctx))


class NarratorTag(BaseTag):
    """Handles > "text" — narrator line, no name box."""
    name = "dialogue.narrator"

    def parse(self, line, line_idx, ctx):
        if not line.startswith(">"):
            return TagParseResult(consumed=False)
        return TagParseResult(
            step={
                "type": "dialogue",
                "speaker_mode": "narrator",
                "raw_text": clean_dialogue_text(line[1:].strip()), # Ensured field consistency
            },
            consumed=True,
        )

    def execute(self, step, ctx):
        if step.get("type") != "dialogue" or step.get("speaker_mode") != "narrator":
            return None
        return ("frame", _build_frame(step, None, ctx))


def _build_frame(step, name, ctx):
    """
    Build the final streamlined dialogue frame for the client.
    Performs standard dynamic python template evaluation and template variable interpolation.
    """
    env = ctx["env"]
    resolved_text = step["raw_text"]

    if resolved_text:
        try:
            # Flatten environmental state dictionary variables map safely into standard .format()
            formatting_map = {}
            for key, value in env.items():
                if key != "Character" and not isinstance(value, type):
                    # If the property resolves to an integer float value, clean the decimal point for display
                    if isinstance(value, float) and value.is_integer():
                        formatting_map[key] = int(value)
                    else:
                        formatting_map[key] = value

            resolved_text = resolved_text.format(**formatting_map)
        except Exception:
            # Fallback strategy: if lookup templates break, keep text untouched to avoid blank screen states
            pass

    return {
        "type": "dialogue",
        "name": name,
        "text": resolved_text, # Returns a plain clean string directly
        "bg": ctx["session"].get("_pending_bg"),
    }

File: src/lib/server/tags/dialogue.ts
import { BaseTag, TagParseResult, type ExecResult } from './base.js';
import type { Step, ExecContext, ParseContext, Frame } from '../types.js';
import { cleanDialogueText } from '../text_utils.js';
import { Character } from '../runtime_types.js';

/**
 * Builds the final dialogue frame, resolving {name} and {obj.field}
 * placeholders against the runtime env. Animated forms ({0..500 : 2000})
 * are left in the text and resolved by the client.
 */
function buildFrame(step: Step, name: string | null, ctx: ExecContext): Frame {
    const raw = step.raw_text as string;
    let resolved = raw;

    if (resolved) {
        resolved = resolved.replace(
            /\{([A-Za-z_][A-Za-z0-9_]*(?:\.[A-Za-z_][A-Za-z0-9_]*)*)\}/g,
            (_match, expr: string) => {
                try {
                    const keys = Object.keys(ctx.env).filter(k =>
                        /^[A-Za-z_$][A-Za-z0-9_$]*$/.test(k)
                    );
                    const values = keys.map(k => ctx.env[k]);
                    // eslint-disable-next-line @typescript-eslint/no-implied-eval
                    const fn = new Function(...keys, `return (${expr});`);
                    const value = fn(...values);
                    if (typeof value === 'number' && Number.isInteger(value)) {
                        return String(value);
                    }
                    return String(value);
                } catch {
                    return `{${expr}}`;
                }
            }
        );
    }

    return {
        type: 'dialogue',
        name,
        text: resolved,
    };
}

export class VariableSpeakerTag extends BaseTag {
    name = 'dialogue.variable';
    private readonly PATTERN = /^:([A-Za-z_][A-Za-z0-9_]*):\s*>\s*(.*)$/;

    parse(line: string, _lineIdx: number, _ctx: ParseContext): TagParseResult {
        const m = this.PATTERN.exec(line);
        if (!m) return new TagParseResult({ consumed: false });
        return new TagParseResult({
            step: {
                type: 'dialogue',
                speaker_mode: 'variable',
                key: m[1],
                raw_text: cleanDialogueText(m[2]),
            },
            consumed: true,
        });
    }

    execute(step: Step, ctx: ExecContext): ExecResult {
        if (step.type !== 'dialogue' || step.speaker_mode !== 'variable') return null;
        const key = step.key as string;
        const obj = ctx.env[key];
        const name = obj instanceof Character ? obj.name : key;
        return ['frame', buildFrame(step, name, ctx)];
    }
}

export class LiteralSpeakerTag extends BaseTag {
    name = 'dialogue.literal';
    private readonly PATTERN = /^([^>]+)>\s*(.*)$/;

    parse(line: string, _lineIdx: number, _ctx: ParseContext): TagParseResult {
        if (line.startsWith('>')) return new TagParseResult({ consumed: false });
        const m = this.PATTERN.exec(line);
        if (!m) return new TagParseResult({ consumed: false });
        return new TagParseResult({
            step: {
                type: 'dialogue',
                speaker_mode: 'literal',
                name: m[1].trim(),
                raw_text: cleanDialogueText(m[2]),
            },
            consumed: true,
        });
    }

    execute(step: Step, ctx: ExecContext): ExecResult {
        if (step.type !== 'dialogue' || step.speaker_mode !== 'literal') return null;
        return ['frame', buildFrame(step, step.name as string, ctx)];
    }
}

export class NarratorTag extends BaseTag {
    name = 'dialogue.narrator';

    parse(line: string, _lineIdx: number, _ctx: ParseContext): TagParseResult {
        if (!line.startsWith('>')) return new TagParseResult({ consumed: false });
        return new TagParseResult({
            step: {
                type: 'dialogue',
                speaker_mode: 'narrator',
                raw_text: cleanDialogueText(line.slice(1).trim()),
            },
            consumed: true,
        });
    }

    execute(step: Step, ctx: ExecContext): ExecResult {
        if (step.type !== 'dialogue' || step.speaker_mode !== 'narrator') return null;
        return ['frame', buildFrame(step, null, ctx)];
    }
}